framework/db/ActiveQuery.php

```diff
@@ -789,4 +800,69 @@ public function alias($alias)
         }
         return $this;
     }
+
+    /**
+     * Replace occurence of self::ALIAS_PLACEHOLDER by $aliasName into given conditions
+     * @param string|array $condition the given conditions
+     * @param string $aliasName
+     */
+    private function replaceAliasPlaceholder($condition, $aliasName)
+    {
+        if ((empty($condition)) or (is_int($condition))) {
+            return $condition;
+        } elseif (is_string($condition)) {
+            return str_replace( self::ALIAS_PLACEHOLDER, $aliasName, $condition);
+        } elseif (is_array($condition)) {
+            $conditionNew = [];
+            foreach ($condition as $key => $value) {
+                unset($condition[$key]);
+                $key    = self::replaceAliasPlaceholder($key,   $aliasName);
+                $value  = self::replaceAliasPlaceholder($value, $aliasName);
+                $conditionNew[$key] = $value;
+            }
+            $condition = $conditionNew;
+        } elseif ($condition instanceof \yii\db\Expression) {
+            $condition->expression = self::replaceAliasPlaceholder($condition->expression, $aliasName);
+        }
+        return $condition;
+    }
+
+    /**
+     * Returns the table alias (or the table name if empty) of the current ActiveQuery
+     * @return string table alias or table name of the current activeQuery
+     */
+    public function getAliasTableName()
+    {
+        list (, $alias) = self::getQueryTableName($this);
+
+        return $alias;
+    }
+
+    /**
+     * Replace the dynamic alias placeholders by the alias table name of the current ActiveQuery
+     * @return $this the current ActiveQuery with relation alias replaced
+     */
+    private function normalizeAliasInQuery()
+    {
+        $alias = $this->getAliasTableName();
+
+        $this->on       = self::replaceAliasPlaceholder( $this->on, $alias );
+        $this->where    = self::replaceAliasPlaceholder( $this->where, $alias );
+        $this->having   = self::replaceAliasPlaceholder( $this->having, $alias );
+        $this->orderBy  = self::replaceAliasPlaceholder( $this->orderBy, $alias );
+        $this->groupBy  = self::replaceAliasPlaceholder( $this->groupBy, $alias );
```

Comment:
Would be nice indeed. I had an unexpected behavior with select during development, I need to check if it's now ok with this last version.